### ISSUE TITLE
Cost planner prefers 2 NodeIndexSeek and CartesianProduct to Expand(All

### ISSUE THREAD

**Issue by frant-hartm** [neo4j/neo4j#12225]
The cost planner fails to produce viable query plan on following query:

```
MATCH (k:Keyword)-[r:CONTAINS_KEYWORD]-(t:Tweet)
WHERE k.value IN [{keywords}] AND t.created_date > {createdDate}
RETURN t,k
```

where there are indexes on Keyword(value) and Tweet(created_date)

**Neo4j Version:** 3.5.5 (both community and enterprise)
**Operating System:** Ubuntu 18.04
**API:** Browser

### Steps to reproduce

Create indexes on 
```
CREATE INDEX ON :Tweet(created_date)
CREATE INDEX ON :Keyword(value)
```

Generate some random data
```
unwind range (1, 100000) as i
create (t:Tweet {created_date:timestamp()})
with t
unwind range(1,10) as j
create (k:Keyword {value:randomUUID()})
create (t)-[:CONTAINS_KEYWORD]->(k)
```
Get some of the generated values, pass to followoing query
```
MATCH (t:Tweet)--(k:Keyword)
RETURN t.created_date-1,k.value LIMIT 1
```

See the query profile:
```
PROFILE
MATCH (k:Keyword)-[r:CONTAINS_KEYWORD]-(t:Tweet)
WHERE k.value IN ['a72e0d93-5445-444c-a022-455d7958e0a6'] AND t.created_date > 1559128418010
return t,k
```
![Query plan](https://imgur.com/download/TvQfRJm)

It picks the wrong index to start with, fair enough. Let's provide the index hint.

```
PROFILE
MATCH (k:Keyword)-[r:CONTAINS_KEYWORD]-(t:Tweet)
USING INDEX k:Keyword(value)
WHERE k.value IN ["25563f3e-8a89-46d5-bacb-ce8fce43cda2"] AND t.created_date > 1559197470542
RETURN t,k
```
![Query plan](https://imgur.com/download/wKPgpvK)

Now it uses both indexes and does cartesian product, bad.

I had to resort to following to trick the planner to execute the desired query plan, note that using only `WITH k,t` doesn't work, need to alias the `t.created_date as cd`

```
PROFILE
MATCH (k:Keyword)-[r:CONTAINS_KEYWORD]-(t:Tweet)
USING INDEX k:Keyword(value)
WHERE k.value IN ["25563f3e-8a89-46d5-bacb-ce8fce43cda2"]
WITH k,t,t.created_date as cd
WHERE cd > 1559197470542
RETURN t,k
```
![Query plan](https://imgur.com/download/QXGjEHT)

## Expected behaviour

Produce the last query plan without any hints, or with `USING INDEX k:Keyword(value)` hint only.

The currently generated query plan might be suitable in certain cases, it should be achievable by using 2 index hints:
```
PROFILE
MATCH (k:Keyword)-[r:CONTAINS_KEYWORD]-(t:Tweet)
USING INDEX k:Keyword(value)
USING INDEX t:Tweet(created_date)
WHERE k.value IN ["25563f3e-8a89-46d5-bacb-ce8fce43cda2"] AND t.created_date > 1559197470542
return t,k
```

**Comment by Hunterness** [comment]
Hi, Thanks for reporting this. We will look into it.
/Team Cypher

**Comment by Lojjs** [comment]
@frant-hartm When looking at your plans we can see that the estimated number of rows for the NodeIndexSeekByRange over Tweet(created_date) is 0. That could indicate that the index statistics have not been updated for some reason. Could you try to add `CALL db.prepareForReplanning` after your data generation to help us confirm that this is the issue?

The procedure will resample the indexes and clear query caches to make sure that the latest index statistics are used.

**Comment by Lojjs** [comment]
@frant-hartm Since we have not heard back from you about the additional information we asked for 2 weeks ago, I will close this issue. If you still have problems, you are very welcome to re-open it with further information included or create a new issue.

**Comment by luanne** [comment]
@Lojjs I'm investigating this at the moment. I could not find a `db.prepareForReplanning` so instead did both `call db.resampleOutdatedIndexes()`, `call db.resampleIndex(":Keyword(value)")`
and `call db.resampleIndex(":Tweet(created_date)")`.

The issue persists. To be sure that the `Keyword` index was being picked up, I dropped the `Tweet` index and profiled the query- works as expected by doing a NodeIndexSeek on the Keyword index and filters down to the 1 row. 
Then I recreated the `Tweet` index on `created_date` and the planner immediately picks up this index to do a NodeIndexSeekByRange instead of picking the more selective Keyword index. Note that estimated rows is still at 0, which is probably why it appears to be the most selective when it's not.

I'm unable to reopen this issue on behalf of Frantisek. Let me know if you prefer a new issue, or whether you will re-open.
BTW I tried this on Neo4j Enterprise 3.5.7

**Comment by fickludd** [comment]
@luanne Thank you for trying all those things.

I am investigating this issue now, and the problem turned out to be caused by two surprising things in combination. 

1) Are you aware that with this setup, all `Tweet`s get the same `created_date`, because `timestamp()` uses the query/statement clock? If you want each `Tweet` to get a slightly different `created_date`, use `datetime.realtime().epochMillis` instead. This also fixed the planning problem.
2) The selectivity estimator for range seeks does not behave nicely for indexes with very low selectivity (like here with only one unique value). In fact when there is only one unique value it believed range scans would never return any rows. That will be fixed in `3.5.9`.

Also note that the `db.resample*` procedures are asynchronous, so it might take some time after that call until the resampling is actually completed. For this setup, you could populate the graph first, and then create indexes, which would ensure sampling was correct immediately when coming online.

**Comment by luanne** [comment]
@fickludd thanks for investigating. The `timestamp()` is indeed something I was not aware of, thanks for pointing that out. 
Good to know about the fix for the index selectivity in 3.5.9!

**Comment by luanne** [comment]
@fickludd sorry, I have a few more observations. Assuming that we did indeed have many tweets with exactly the same timestamp, providing the index hint only partially solves the issue.

In my test database, the keyword is related to exactly one tweet which matches the date filter:
![image](https://user-images.githubusercontent.com/1242476/63223578-db659b80-c1c8-11e9-9265-601e295b2dd9.png)

So, I'd expect that the expansion happens via the single keyword to the single tweet but the plan for this query with the index hint
```
PROFILE
MATCH (k:Keyword)<-[r:CONTAINS_KEYWORD]-(t:Tweet)
USING INDEX k:Keyword(value)
WHERE k.value IN ['c4160609-ebf9-4197-9dc3-86839e3bd278'] AND t.created_date > 1566123035893
return t,k
```

is 
![image](https://user-images.githubusercontent.com/1242476/63223696-fafdc380-c1ca-11e9-835f-fcad244b2678.png)

And now for a bit of fun ;-)
If you flip the order of the MATCH expression:

```
PROFILE
MATCH (t:Tweet)-[r:CONTAINS_KEYWORD]->(k:Keyword)
USING INDEX k:Keyword(value)
WHERE k.value IN ['c4160609-ebf9-4197-9dc3-86839e3bd278'] AND t.created_date > 1566123035893
return t,k
```

the query plan changes to

![image](https://user-images.githubusercontent.com/1242476/63223713-1c5eaf80-c1cb-11e9-9180-72453030208f.png)

:-)

**Comment by sherfert** [comment]
@luanne I expect both of these plans to change in 3.5.9 with the bugfix. Could you verify that when 3.5.9 is out?

**Comment by tomswinkels** [comment]
https://github.com/neo4j/neo4j/issues/12286

**Comment by luanne** [comment]
@sherfert I'm confirming that these are fixed in 3.5.9, thanks!
